### ISSUE TITLE
issues with DSSATResults

### ISSUE THREAD

**Issue by FabioSeixas** [julienmalard/traDSSAT#19]
**Describe the bug**
I found two issues trying to use `DSSATResults`:

The first, when `PlantGro.OUT` is in the Outputs directory:

```
Traceback (most recent call last):                                                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 199, in _get_var_names              
    name = next(vr for vr in var_names if line.startswith(vr))                                                                                          
StopIteration                                                                                                                                           
                                                                                                                                                        
During handling of the above exception, another exception occurred:                                                                                     
                                                                                                                                                        
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 7, in <module>                                                                                                               
    print(x.get_value(var = "WUPRD", trt = 1, t = None, at = "0"))                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\out.py", line 73, in get_value                     
    f = self._outfiles[c] = c(self.folder)                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\output.py", line 23, in __init__                   
    super().__init__(file)                                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 29, in __init__                     
    self._read()                                                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 44, in _read                        
    self._read_section(section)                                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 171, in _read_section               
    self._read_subsection(section_name, subblock)                                                                                                       
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 126, in _read_subsection            
    var_names = self._get_var_names(subblock[0])                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 201, in _get_var_names              
    raise ValueError(                                                                                                                                   
ValueError: No variable matching "TMEAN  GSTD  L#SD PA" for file PlantGro.OUT. 
```

When I remove `PlantGro.OUT` from the Outputs directory and try **a lot of variables** in `get_value()`, the same error appears:

```
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 7, in <module>                                                                                                               
    print(x.get_value(var = "WUPRD", trt = 1, t = None, at = "0"))                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\out.py", line 77, in get_value                     
    if var in f.variables():                                                                                                                            
AttributeError: 'NoneType' object has no attribute 'variables'
```

**To Reproduce**

```
import tradssat
tradssat.set_dssat_dir("C:/DSSAT47")
x = tradssat.DSSATResults("C:/DSSAT47/Cassava")
print(x.get_value(var = "WUPRD", trt = 1, t = None, at = "0"))
```

2. Version of traDSSAT: 0.1.5

3. Output files:
[PlantGrf.txt](https://github.com/julienmalard/traDSSAT/files/4021320/PlantGrf.txt)
[ET.txt](https://github.com/julienmalard/traDSSAT/files/4021321/ET.txt)
[Weather.txt](https://github.com/julienmalard/traDSSAT/files/4021322/Weather.txt)
[SoilTemp.txt](https://github.com/julienmalard/traDSSAT/files/4021323/SoilTemp.txt)
[PlantGro.txt](https://github.com/julienmalard/traDSSAT/files/4021324/PlantGro.txt)
[SoilWat.txt](https://github.com/julienmalard/traDSSAT/files/4021325/SoilWat.txt)

**Comment by julienmalard** [comment]
Hello,

Sorry for the delay! I hope it is not too late for your work. I checked the code and the main problem is that most of these variables were missing from traDSSAT. Unfortunately, it seems that different models' outputs have columns with different widths *for the same variables*. This means that we will have to tweak traDSSAT internals to delineate columns differently, perhaps according to the alignment in the header row.

I will try to get around to this soon! In the meantime, if you think of a solution, please feel free to contribute a pull request as well.
Best,
Julien

**Comment by FabioSeixas** [comment]
Hello Julien,

Hope you are doing well and stay safe.

Yes, It's very clear that the same rule for each variable will not work for all crops. The actual rule works fine for Pineapple, and simply not work for Maize and Cassava (the only crops I tested). 

------------------------------
> This means that we will have to tweak traDSSAT internals to delineate columns differently, perhaps according to the alignment in the header row.

I was thinking, currently rule uses `spc = 1` for all variables and `size` is according to Pineapple output. Maybe if `size = len(var_name)` and `spc` is computed reading the space strings within the `_read_subsection` method, a more flexible rule can be obtained.

The header can be used to compute the spaces before the column title, and then the cutoffs can be computed. 

An additional problem that I found on Cassava `PlantGro.OUT` was between the columns `HWAD ` and `HIAD`. You need to see that by yourself ([PlantGro.txt](https://github.com/julienmalard/traDSSAT/files/4604384/PlantGro.txt)). It's not present in the previous PlantGro.txt I sent to you. I think it is a bug at the new crop version. Honestly, the best solution I see is to cut this variable out.

Well, I'm willing to try to implement the solution, if it seems reasonable to you. If it's not, please share you with me your thoughts.

**Comment by FabioSeixas** [comment]
Hello again,

I worked on the above solution and, as you should have imagined, it broke other files processing (I got **7 Fails** and **186 Errors** running `test_file.py`). 

I don't know what to do, but I'm quite sure that the safer way is to do something on `PlantGro` or even on `OutFile` classes. Otherwise, it can crash other files processing.

I will wait your answer to continue.

Best!

**Comment by julienmalard** [comment]
Hello,

I think I just answered most issues here in my response to your other email. Regarding the HWAD-HIAD, though, that is impressive! One idea would be to get the size of each variable only from the size of its name in the header, which is what I think your pull request does - so in this case perhaps it would work with HWAD and HIAD? I am also considering changing traDSSAT so that it can read all variables, including ones it does not know about, which would require a bit more coding work but would fix all those annoying I-forgot-about-that-variable bugs!

Thanks a lot,
Julien
________________________________
દ્વારા: Fábio Seixas <notifications@github.com>
મોકલ્યું: 09 મે 2020 22:15
પ્રતિ: julienmalard/traDSSAT <traDSSAT@noreply.github.com>
Cc: Julien Malard <julien.malard@mail.mcgill.ca>; Comment <comment@noreply.github.com>
વિષય: Re: [julienmalard/traDSSAT] issues with DSSATResults (#19)


Hello again,

I worked on the above solution and, as you should have imagined, it broke other files processing (I got 7 Fails and 186 Errors running test_file.py).

I don't know what to do, but I'm quite sure that the safer way is to do something on PlantGro or even on OutFile classes. Otherwise, it can crash other files processing.

I will wait your answer to continue.

Best!

**Comment by FabioSeixas** [comment]
> One idea would be to get the size of each variable only from the size of its name in the header, which is what I think your pull request does - so in this case perhaps it would work with HWAD and HIAD?

Unfortunally not. The only case on which the PR was not successfull was the HWAD-HIAD.

I think the responsible is the following code in `_read_subsection` method. 

```
for i, l in enumerate(subblock[1:]):
            # Odd workaround necessary because several cultivar names in DSSAT are larger than the allowed space
            # and so run into the next column, which apparently isn't supposed to matter if the next column's value
            # is small enough to allow both to fit. (Really?!)
            vals = [
                (l[0 if c[0] == 0 else max(c[0], l.find(' ', c[0], c[1] - 1)):
                   None if l.find(' ', c[1] - 1) < 0 else l.find(' ', c[1] - 1)]).strip()
                for c in cutoffs]
```

It uses the cutoffs to search for the values. It understands a space `" "` as the end of the value number. In the HWAD-HIAD, the end is a `"-"`. 

I tried some changes but always got something broken. 

And I really think It's a bug from the new cassava model version. So, It not worth to waste time on that.

Thanks!

**Comment by julienmalard** [comment]
Yes, I think you are right...

**Comment by julienmalard** [comment]
Closed by #20.
